sources/packages/backend/apps/api/src/services/application-restriction-bypass/application-restriction-bypass.service.ts

```diff
@@ -43,10 +67,382 @@ export class ApplicationRestrictionBypassService extends RecordDataModelService<
       where: {
         application: {
           id: applicationId,
-          applicationStatus: Not(ApplicationStatus.Draft),
         },
       },
       order: { createdAt: "DESC" },
     });
   }
+
+  /**
+   * Gets an application restriction bypass by id.
+   * @param id id of the application restriction bypass to retrieve.
+   * @returns application restriction bypass.
+   */
+  async getApplicationRestrictionBypass(
+    id: number,
+  ): Promise<ApplicationRestrictionBypass> {
+    return this.applicationRestrictionBypassRepo.findOne({
+      select: {
+        id: true,
+        isActive: true,
+        studentRestriction: {
+          id: true,
+          restriction: {
+            restrictionCode: true,
+          },
+        },
+        creationNote: {
+          description: true,
+        },
+        removalNote: {
+          description: true,
+        },
+        bypassCreatedBy: {
+          firstName: true,
+          lastName: true,
+        },
+        bypassCreatedDate: true,
+        bypassRemovedBy: {
+          firstName: true,
+          lastName: true,
+        },
+        bypassRemovedDate: true,
+        bypassBehavior: true,
+      },
+      relations: {
+        studentRestriction: { restriction: true },
+        creationNote: true,
+        removalNote: true,
+        bypassCreatedBy: true,
+        bypassRemovedBy: true,
+      },
+      where: {
+        id,
+      },
+    });
+  }
+
+  /**
+   * Gets all available student restrictions to bypass for a given application.
+   * @param applicationId application id.
+   * @returns list of application restriction bypass options.
+   */
+  async getAvailableStudentRestrictionsToBypass(
```

Comment:
This method now executes three SQL queries to filter an amount of data that should be pretty small (1 to few active records).
My recommendation is to get from DB everything that is needed, as much as possible with a single query and do minor in-memory process to finish the filter.
For instance, the below query would return
- all active student restrictions (expect zero records to couples).
- all active bypasses (expect zero records to couples).
- only restriction with the expected action types.
- offering intensity.
From the returned query, we can eliminate the ones already in place and the ones for the action types not needed for the offering intensity.
```ts
this.applicationRepo.find({
  select: {
    id: true,
    student: {
      id: true,
      studentRestrictions: {
        id: true,
        restriction: {
          id: true,
          restrictionCode: true,
        },
        restrictionBypasses: {
          id: true,
          studentRestriction: { id: true },
        },
      },
    },
    currentAssessment: {
      id: true,
      offering: {
        id: true,
        offeringIntensity: true,
      },
    },
  },
  relations: {
    student: {
      studentRestrictions: { restriction: true, restrictionBypasses: true },
    },
    currentAssessment: { offering: true },
  },
  where: {
    id: applicationId,
    student: {
      studentRestrictions: {
        isActive: true,
        restrictionBypasses: {
          isActive: true,
        },
        restriction: {
          actionType: ArrayContains([
            RestrictionActionType.StopFullTimeBCFunding,
            RestrictionActionType.StopFullTimeDisbursement,
            RestrictionActionType.StopPartTimeBCFunding,
            RestrictionActionType.StopPartTimeDisbursement,
          ]),
        },
      },
    },
  },
});
```

Reply:
I agree with this approach to boost performance for database querying.

Reply:
Changes looks great, thanks for making them happen.
As a suggestion, I would reoder and drop some comments as below. Not a blocker.
```ts
// Filter all active bypasses.
    const bypassedStudentRestrictionIds =
      application.student.studentRestrictions
        .flatMap(
          (studentRestriction) =>
            studentRestriction.applicationRestrictionBypasses,
        )
        .filter((bypass) => bypass.isActive)
        .map((bypass) => bypass.studentRestriction.id);
    // Filter all active student restriction that does not have an active bypass.
    const studentRestrictionsThatAreNotBypassed =
      application.student.studentRestrictions.filter(
        (studentRestriction) =>
          !bypassedStudentRestrictionIds.includes(studentRestriction.id),
      );
    // Define the specific restrictions actions allowed for the application,
    // to filter the returned student restrictions for the requested application.
    const allowedRestrictionActions =
      application.currentAssessment.offering.offeringIntensity ===
      OfferingIntensity.fullTime
        ? [
            RestrictionActionType.StopFullTimeBCFunding,
            RestrictionActionType.StopFullTimeDisbursement,
          ]
        : [RestrictionActionType.StopPartTimeDisbursement];
    return studentRestrictionsThatAreNotBypassed
      .filter((studentRestriction) =>
        allowedRestrictionActions.some((actionType) =>
          studentRestriction.restriction.actionType.includes(actionType),
        ),
      )
      .map((studentRestriction) => ({
        studentRestrictionId: studentRestriction.id,
        restrictionCode: studentRestriction.restriction.restrictionCode,
        studentRestrictionCreatedAt: studentRestriction.createdAt,
      }))
      .sort((a, b) => a.restrictionCode.localeCompare(b.restrictionCode));
```